src/test/java/spoon/test/template/testclasses/match/MatchMultiple.java

```diff
@@ -0,0 +1,55 @@
+package spoon.test.template.testclasses.match;
+
+import spoon.pattern.Pattern;
+import spoon.pattern.PatternBuilder;
+import spoon.pattern.PatternBuilderHelper;
+import spoon.pattern.matcher.Quantifier;
+import spoon.reflect.code.CtLiteral;
+import spoon.reflect.declaration.CtType;
+import spoon.reflect.meta.ContainerKind;
+import spoon.testing.utils.ModelUtils;
+
+import static java.lang.System.out;
+
+public class MatchMultiple {
+
+	/** return a pattern built from {}@link {@link #matcher1()} */
+	public static Pattern createPattern(Quantifier matchingStrategy, Integer minCount, Integer maxCount) throws Exception {
+		CtType<?> type = ModelUtils.buildClass(MatchMultiple.class);
+		return PatternBuilder.create(new PatternBuilderHelper(type).setBodyOfMethod("matcher1").getPatternElements())
+			.configureParameters(pb -> {
+
+				// matching anything that is called "statements".
+				// the setContainerKind(ContainerKind.LIST) means match zero, one or more then one arbitrary statement
+				pb.parameter("statements").byReferenceName("statements").setContainerKind(ContainerKind.LIST);

```

Comment:
I see, what I don't understand is that the reference is matched, which means the executable ref in this case.
while the pattern is actually the CtInvocation, ie the parent of the executable reference.

When does the link to parent is made? Is it automated?

Reply:
Yes, it is automated by this code:
/spoon-core/src/main/java/spoon/pattern/ParametersBuilder.java (line 629)
```java
	public ParameterElementPair getSubstitutedNodeOfElement(ParameterInfo parameter, CtElement element) {
		ParameterElementPair parameterElementPair = new ParameterElementPair(parameter, element);
		parameterElementPair = transformVariableAccessToVariableReference(parameterElementPair);
		parameterElementPair = transformArrayAccess(parameterElementPair);
		parameterElementPair = transformTemplateParameterInvocationOfS(parameterElementPair);
		parameterElementPair = transformExecutableRefToInvocation(parameterElementPair);
		parameterElementPair = transformCtReturnIfNeeded(parameterElementPair);
		//if spoon creates an implicit parent (e.g. CtBlock) around the pattern parameter, then replace that implicit parent
		parameterElementPair = getLastImplicitParent(parameterElementPair);
		return parameterElementPair;
	}
```